dtfs-central-api/src/v1/controllers/utilisation-report-service/post-keying-data.controller/index.ts

```diff
@@ -1,24 +1,50 @@
-import { Request, Response } from 'express';
+import { Response } from 'express';
 import { HttpStatusCode } from 'axios';
-import { ApiError } from '@ukef/dtfs2-common';
-import { UtilisationReportRepo } from '../../../../repositories/utilisation-reports-repo';
-import { NoMatchingFeeRecordsError, NotFoundError } from '../../../../errors';
+import { ApiError, CustomExpressRequest } from '@ukef/dtfs2-common';
+import { FeeRecordRepo } from '../../../../repositories/fee-record-repo';
+import { NotFoundError } from '../../../../errors';
+import { executeWithSqlTransaction } from '../../../../helpers';
+import { UtilisationReportStateMachine } from '../../../../services/state-machines/utilisation-report/utilisation-report.state-machine';
+import { getGenerateKeyingDataDetails } from './helpers';
+import { PostKeyingDataPayload } from '../../../routes/middleware/payload-validation';
 
-export const postKeyingData = async (req: Request, res: Response) => {
+export type PostKeyingDataRequest = CustomExpressRequest<{
+  reqBody: PostKeyingDataPayload;
+}>;
+
+export const postKeyingData = async (req: PostKeyingDataRequest, res: Response) => {
   const { reportId } = req.params;
+  const { user } = req.body;
 
   try {
-    const utilisationReport = await UtilisationReportRepo.findOneByIdWithFeeRecordsFilteredByStatus(Number(reportId), ['MATCH']);
+    const feeRecordsWhichCanBeKeyed = await FeeRecordRepo.findByReportIdAndStatusesWithReport(Number(reportId), ['MATCH']);
 
-    if (!utilisationReport) {
-      throw new NotFoundError(`Failed to find a report with id '${reportId}'`);
+    if (feeRecordsWhichCanBeKeyed.length === 0) {
+      throw new NotFoundError(`Failed to find any fee records which can be keyed attached to report with id '${reportId}'`);
     }
 
-    const { feeRecords: matchingFeeRecords } = utilisationReport;
-    if (matchingFeeRecords.length === 0) {
-      throw new NoMatchingFeeRecordsError();
+    const utilisationReport = feeRecordsWhichCanBeKeyed[0].report;
+    if (!utilisationReport) {
+      throw new Error('The report on the found fee records is not defined');
     }
 
+    const generateKeyingDataDetails = await getGenerateKeyingDataDetails(feeRecordsWhichCanBeKeyed);
```

Comment:
I feel like this is a little opaque in terms of the naming since `GenerateKeyingDataDetails` doesn't really tell us what it is until we dig down into, I tried to think of a better name but failed but realised I almost expect this to happen inside the event handler? Then we could just pass the utilisationReportStateMachine a list of FeeRecords and at least contain this slightly odd type to where it is crucial?